README.md
# libunwind Package Recipe for the conan Package Manager

This is an effort to make [libunwind]
available through the [conan Package Manager].

[![Build Status](https://travis-ci.org/MarcoKoch/conan-libunwind.svg?branch=master)](https://travis-ci.org/MarcoKoch/conan-libunwind) ([master], upcoming version)  
[![Build Status](https://travis-ci.org/MarcoKoch/conan-libunwind.svg?branch=v1.2.1)](https://travis-ci.org/MarcoKoch/conan-libunwind) ([v1.2.1], current version)

**NOTE:**
This repository only contains the source code of the conan package recipe.
This is *not* the original libunwind source code
and this repository is not maintained by the libunwind developer(s).
The original source code is downloaded automatically
from the official libunwind website
during the package build process.
Please visit libunwind's website at <http://www.nongnu.org/libunwind>.


## Hosted Packages

Recipes and binary packages built from this repository
are hosted at the [Ocean Conan Package Repository].
To use the repository, add it to your conan remotes like this:

    conan remote add ocean-conan https://api.bintray.com/conan/marcokoch/ocean-conan

The packages are available in two channels:

`libunwind/1.2.1@marcokoch/stable`:
This channel contains the latest stable package release, built
from the [v1.2.1-0] tag in this repository.
Note that 'stable' refers to the packaging only. The package is always built
from stable releases of libunwind.

`libunwind/1.2.1@marcokoch/testing`:
This channel provides recent development builds from the [v1.2.1]
branch of this repository.
Feedback on these builds is highly welcome.
They may be buggy, though.

To use libunwind in your project,
use the following lines in your `conanfile.txt`:

    [requires]
    libunwind/1.2.1@marcokoch/stable

Or in your `conanfile.py`:

    class MyProject(ConanFile):
        requires = "libunwind/1.2.1@marcokoch/stable"
        
There are binary packages available for the following configurations:

| Operating System | Architectures | Compilers              | `build_type`       | `shared`        |
|------------------|---------------|------------------------|--------------------|-----------------|
| Linux            | i686, x86_64  | GCC 5.2, 5.3, 5.4, 6.3 | "Debug", "Release" | `True`, `False` |
| Linux            | i686, x86_64  | Clang 3.8, 3.9, 4.0    | "Debug", "Release" | `True`, `False` |
        
        
## Platform support

This recipe should build for all platforms
that are supported by both libunwind and conan.
At the moment, these are:

* Linux (x86)
* Linux (x86_64)
* Linux (ARM)
* Linux (PPC64)
* MIPS
* FreeBSD (x86)
* FreeBSD (x86_64)

If a package build breaks for any of those platforms, please report a bug on
the [issue tracker].

**NOTE:**
Unfortunately, there is no support for Windows and/or OS X at the moment.


## Compiler support

Packages from this recipe can both be compiled and used with

* GCC versions 5.1, 5.2, 5.3, 5.4, 6.1, 6.2, 6.3, 7.1
* Clang versions 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 5.0


## Package Options

The package supports the following options:

| Option                       | Possible Values | Default | Description                                       | Equivalent libunwind configure script option |
|------------------------------|-----------------|---------|---------------------------------------------------|----------------------------------------------|
| `shared`                     | `True`, `False` | `False` | Build shared libraries                            | `--enable-shared`                            |
| `enable_coredump`            | `True`, `False` | `False` | Include the unwind-coredump library               | `--enable-coredump`                          |
| `enable_ptrace`              | `True`, `False` | `False` | Include the unwind-ptrace library                 | `--enable-ptrace`                            |
| `enable_setjmp`              | `True`, `False` | `False` | Include the unwind-setjmp library                 | `--enable-setjmp`                            |
| `enable_cxx_exceptions`      | `True`, `False` | `False` | Use libunwind to handle C++ exceptions            | `--enable-cxx-exceptions`                    |
| `enable_debugframe`          | `True`, `False` | `False` | Load the ".debug_frame" section if available      | `--enable-debugframe`                        |
| `enable_block_signals`       | `True`, `False` | `False` | Block signals before performing mutext operations | `--enable-block-signals`                     |
| `enable_conservative_checks` | `True`, `False` | `False` | Validate all memory addresses before use          | `--enable-conservative-checks`               |
| `enable_msabi_support`       | `True`, `False` | `False` | Enable support for Microsoft ABI extensions       | `--enable-msabi-support`                     |
| `enable_minidebuginfo`       | `True`, `False` | `False` | Enable support for LZMA-compressed symbol tables  | `--enable-minidebuginfo`                     |


## License

This package recipe is provided under the 3-Clause BSD license.
See file [LICENSE](LICENSE) for further information.

The original libunwind project is provided under the MIT license.
Please see <https://github.com/libunwind/libunwind/blob/v1.2-stable/LICENSE>.


## Contributing

Feedback and pull requests are highly appreciated.
If you have any questions, feel free to ask in the [issue tracker].

To contribute to the original libunwind project,
please interact with the official GitHub mirror at <https://github.com/libunwind/libunwind>.


[conan package manager]: https://conan.io/
[libunwind]: http://www.nongnu.org/libunwind/
[Ocean Conan Package Repository]: https://bintray.com/marcokoch/ocean-conan
[issue tracker]: https://github.com/MarcoKoch/conan-libunwind/issues
[libunwind README]: https://github.com/libunwind/libunwind/blob/master/README
[v1.2.1]: https://github.com/MarcoKoch/conan-libunwind/tree/v1.2.1
[v1.2.1-0]: https://github.com/MarcoKoch/conan-libunwind/tree/v1.2.1-0
[master]: https://github.com/MarcoKoch/conan-libunwind/tree/master


CHANGES
Version 1.2.1-0 (2017-07-01)

    Initial release


conanfile.py
from conans import ConanFile, AutoToolsBuildEnvironment, tools
import os, re


class LibunwindConan(ConanFile):
    name = "libunwind"
    version = "1.2.1"
    license = "MIT"
    url = "https://github.com/marcokoch/conan-libunwind"
    author = "Marco Koch (marco-koch@t-online.de)"
    description = "A complete open-source implementation of the libunwind API "" \
""currently exists for IA-64 Linux and is under development for Linux on x86, "" \
""x86-64, and PPC64. Partial support for HP-UX and Linux on PA-RISC also exists."
    settings = {
"os" : ["Linux", "FreeBSD"],
"compiler": {
    "gcc" : {
        "version": ["5.1", "5.2", "5.3", "5.4", "6.1", "6.2", "6.3"],
        "libcxx": None,
        "threads": None,
        "exception": None
    },
    "clang": None
},
"build_type": None,
"arch": ["x86", "x86_64", "ppc64le", "ppc64", "armv6", "armv7", "armv7hf", "armv8", "mips", "mips64"]}
    options = {
"shared": [True, False],
"enable_coredump": [True, False],
"enable_ptrace": [True, False],
"enable_setjmp": [True, False],
# Not supported for now, as we don not have a latex2man package
#"enable_documentation": [True, False],
"enable_cxx_exceptions": [True, False],
"enable_debugframe": [True, False],
"enable_block_signals": [True, False],
"enable_conservative_checks": [True, False],
"enable_msabi_support": [True, False],
"enable_minidebuginfo": [True, False]
    }
    default_options = '''
shared=False
enable_coredump=False
enable_ptrace=False
enable_setjmp=False
enable_cxx_exceptions=False
enable_debugframe=False
enable_block_signals=False
enable_conservative_checks=False
enable_msabi_support=False
enable_minidebuginfo=False
'''

    source_archive_name = "libunwind-%s" % version
    source_url = "http://download.savannah.nongnu.org/releases/libunwind/%s.tar.gz" % source_archive_name
    local_install_path = "install"

    
    def _get_target_environment_str(self):
        if self.settings.arch == "x86":
            arch = "i686"
        else:
            arch = self.settings.arch
        
        if self.settings.os == "Linux":
            return "%s-linux" % arch
        elif self.settings.os == "FreeBSD":
            return "%s-freebsd" % arch
            
            
    def configure(self):
        if not re.search("^mips(64)?$", "%s" % self.settings.arch) \
                and not ((self.settings.os == "Linux" \
                        and re.search("^(x86(_64)?)|(armv[0-9](hf)?)|(ppc64(le)?)$", "%s" % self.settings.arch)) \
                    or (self.settings.os == "FreeBSD" \
                        and re.search("^x86(_64)?$", "%s" % self.settings.arch))):
            raise Exception( \
                "Your chosen architecture (%s) is not supported by libunwind on this operating system (%s)" \
                    % (self.settings.arch, self.settings.os))


    def source(self):
        self.output.info("Downloading sources from %s" % self.source_url)
        tools.get(self.source_url)


    def build(self):
        build_env = AutoToolsBuildEnvironment(self)        
        configure_options = ["--prefix="]
        
        if self.settings.build_type == "Debug":
            configure_options.append("--enable-debug")
        
        if self.options.shared:
            configure_options.append("--disable-static")
        else:
            configure_options.append("--disable-shared")
        
        if not self.options.enable_coredump:
            configure_options.append("--disable-coredump")
        
        if not self.options.enable_ptrace:
            configure_options.append("--disable-ptrace")
        
        if not self.options.enable_setjmp:
            configure_options.append("--disable-setjmp")
        
        # Not supported for now, as we don not have a latex2man package
        #if not self.options.enable_documentation:
        #    configure_options.append("--disable-documentation")
        
        if self.options.enable_cxx_exceptions:
            configure_options.append("--enable-cxx-exceptions")
        
        if self.options.enable_debugframe:
            configure_options.append("--enable-debugframe")
        
        if self.options.enable_block_signals:
            configure_options.append("--enable-block-signals")
            
        if self.options.enable_conservative_checks:
            configure_options.append("--enable-conservative-checks")
        
        if self.options.enable_msabi_support:
            configure_options.append("--enable-msabi-support")
    
        if self.options.enable_minidebuginfo:
            configure_options.append("--enable-minidebuginfo")
        
        with tools.chdir(self.source_archive_name):
            self.output.info("Configuring libunwind")
            build_env.configure(args=configure_options, host=self._get_target_environment_str())
            
            self.output.info("Building libunwind")
            build_env.make()
            
            self.output.info("Doing a local install for packaging")
            build_env.make(["install", "prefix=%s" % os.path.join(os.getcwd(), \
                os.pardir, self.local_install_path)])
        
        # HACK: There seems to be a bug in libunwind's build system that causes
        # a broken symlink to be created when configured with --enable-shared.
        # This causes conan to drop out with an error when it tries to copy
        # that link during package(). The easiest workaround should be to just
        # delete that nonsense.
        if self.options.shared:
            broken_link = os.path.join(self.local_install_path, "lib", "libunwind-generic.a")
            if os.path.islink(broken_link):
                self.output.info("HACK: Deleting broken symlink %s" % broken_link)
                os.unlink(broken_link)
        

    def package(self):
        self.copy("*", dst=".", src=self.local_install_path, links=True)
        self.copy("COPYING", dst="licenses", src=self.source_archive_name)


    def package_info(self):
        self.cpp_info.libs = ["unwind"]
        
        if self.options.enable_coredump:
            self.cpp_info.libs.append("unwind-coredump")
        
        if self.options.enable_ptrace:
            self.cpp_info.libs.append("unwind-ptrace")
        
        if self.options.enable_setjmp:
            self.cpp_info.libs.append("unwind-setjmp")
